eventsourcing-core/src/main/java/com/eventsourcing/EventStream.java

```diff
@@ -0,0 +1,152 @@
+ */
+public class EventStream<S> {
+    @Getter
+    private Stream<? extends Event> stream;
+    @Getter
+    private S state;
+
+    EventStream(S state, Stream<? extends Event> stream) {
+        this.state = state;
+        this.stream = stream;
+    }
+
+    public static class Builder<S>  {
+
+        private final S state;
+        private final Stream.Builder<Event> builder = Stream.builder();
+
+        public Builder(S state) {
+            this.state = state;
+        }
+
+        public void accept(Event event) {
+            builder.accept(event);
+        }
+
+        public Builder<S> add(Event event) {
+            accept(event);
+            return this;
+        }
+
+        public EventStream<S> build() {
+            return new EventStream<>(state, builder.build());
+        }
+    }
+
+    /**
+     * EventStream builder
+     *
+     * @param state state
+     * @param <S> state type
+     * @return
+     */
+    public static <S> Builder<S> builder(S state) {
+        return new Builder<>(state);
+    }
+
+    /**
+     * EventStream builder with state set to <code>null</code>
+     *
+     * @param <S> state type
+     * @return
+     */
+    public static <S> Builder<S> builder() {
+        return new Builder<>(null);
+    }
+
+    /**
+     * @param <S> state type
+     * @return empty event stream with state set to <code>null</code>
+     */
+    public static <S> EventStream<S> empty() {
+        return new EventStream<>(null, Stream.empty());
+    }
+
+    /**
+     * @param state state
+     * @param <S> state type
+     * @return empty event stream with a state
+     */
+    public static <S> EventStream<S> empty(S state) {
+        return new EventStream<>(state, Stream.empty());
+    }
+
+    /**
+     * @param state state
+     * @param stream stream of events
+     * @param <S> state type
+     * @return event stream with a state and a stream
+     */
+    public static <S> EventStream<S> ofWithState(S state, Stream<? extends Event> stream) {
+        return new EventStream<>(state, stream);
+    }
+
+    /**
+     * @param stream stream of events
+     * @param <S> state type
+     * @return event stream with a state set to <code>null</code> and a stream
+     */
+    public static <S> EventStream<S> of(Stream<? extends Event> stream) {
+        return new EventStream<>(null, stream);
+    }
+
+    /**
+     * @param state state
+     * @param event event
+     * @param <S> state type
+     * @return event stream with a state and a stream of one event
+     */
+    public static <S> EventStream<S> ofWithState(S state, Event event) {
+        return new EventStream<>(state, Stream.of(event));
+    }
+
+    /**
+     * @param event event
+     * @param <S> state type
+     * @return event stream with a state set to <code>null</code> and a stream of one event
+     */
+    public static <S> EventStream<S> of(Event event) {
+        return new EventStream<>(null, Stream.of(event));
+    }
+
+    /**
```

Comment:
Minor nitpick: Pull `ofWithState(S, Event...)` one up for consistency (`ofWithState()`, `of()`, `ofWithState()`, `of()`, ...).